File: algorithms/medium/different_ways_to_add_parentheses.py
from functools import cache
from typing import List

class Solution:
    def diffWaysToCompute(self, expression: str) -> List[int]:
        @cache
        def solve(i, j):
            if expression[i:j+1].isdigit():
                return [int(expression[i:j+1])]
            arr = list()
            for k in range(i, j+1):
                if not expression[k].isdigit():
                    op = expression[k]
                    left_arr = solve(i, k-1)
                    right_arr = solve(k+1, j)
                    for lv in left_arr:
                        for rv in right_arr:
                            if op == '+':
                                arr.append(lv + rv)
                            elif op == '-':
                                arr.append(lv - rv)
                            elif op == '*':
                                arr.append(lv * rv)
            return arr
        N = len(expression)
        i, j = 0, N - 1
        res = solve(i, j)
        return res

# Main section
for expression in [
                     '2-1-1',
                     '2*3-4*5',
                     '2*7+3',
                     '2*3-4+5+6*7-8*9+1',
                     '1+2-3*4+5-6*7+8-9*10',
                     '17+37*53-79+83*91-97',
                     '98',
                     '98-98',
                  ]:
    print(f'expression = {expression}')
    sol = Solution()
    r = sol.diffWaysToCompute(expression)
    print(f'r = {r}')
    print(f'len(r) = {len(r)}')
    print('===================')

File: algorithms/medium/different_ways_to_add_parentheses.go
package main

import (
    "fmt"
    "strconv"
)

func solve(expression string, i, j int) []int {
    n, err := strconv.Atoi(expression[i : j+1])
    if err == nil {
        return []int{n}
    }
    arr := []int{}
    for k := i; k < j+1; k++ {
        n, err = strconv.Atoi(string(expression[k]))
        if err != nil {
            op := expression[k]
            leftArr := solve(expression, i, k-1)
            rightArr := solve(expression, k+1, j)
            for _, lv := range leftArr {
                for _, rv := range rightArr {
                    if op == '+' {
                        arr = append(arr, lv+rv)
                    } else if op == '-' {
                        arr = append(arr, lv-rv)
                    } else if op == '*' {
                        arr = append(arr, lv*rv)
                    }
                }
            }
        }
    }
    return arr
}

func diffWaysToCompute(expression string) []int {
    N := len(expression)
    i, j := 0, N-1
    results := solve(expression, i, j)
    return results
}

func main() {
    data := []string{
        "2-1-1",
        "2*3-4*5",
        "2*7+3",
        "2*3-4+5+6*7-8*9+1",
        "1+2-3*4+5-6*7+8-9*10",
        "17+37*53-79+83*91-97",
        "98",
        "98-98",
    }
    for _, expression := range data {
        r := diffWaysToCompute(expression)
        fmt.Println(expression)
        fmt.Println(r)
        fmt.Println("=======================")
    }
}

